File: packages/scenario/src/schema/v2/set-keys.ts
/**
 * The typed key registry behind the `set` verb.
 *
 * `set` is the discrete-state escape valve — indicators, doors, a flagger's
 * paddle, a signal phase, and the behaviour switches that decide whether a
 * challenger actually commits. Left open it becomes a stringly-typed dumping
 * ground that no engine can implement completely; so the legal keys are **data**
 * (this file), each with a value type, an axis, and the actor kinds it applies
 * to, and anything else is rejected with the near-miss suggestion.
 *
 * The most important entry is `rules.collisionAvoidance`. Per
 * `docs/research/interactions-and-edge-cases.md` §three make-or-break features,
 * without the ability to turn it off, every generated critical scenario
 * silently degrades into a safe one because the challenger chickens out.
 *
 * Wildcard keys (`signal:<handle>.phase`) are matched by pattern, since the
 * handles come from the map, not from the schema.
 */

import { z } from 'zod';

/** Value domain of a settable key. */
export const SetValueTypeSchema = z.enum(['boolean', 'number', 'enum', 'string']);

/** Who a key may be set on. */
export const SetTargetKindSchema = z.enum(['vehicle', 'vru', 'any_actor', 'world']);

/** One registry entry. */
export const SetKeyDeclSchema = z.strictObject({
  /** Literal key, or the human form of `pattern` when one is present. */
  key: z.string().min(1).max(100),
  /**
   * Regex source for wildcard keys (`signal:<handle>.phase`). When present,
   * `key` is documentation and matching goes through the pattern.
   */
  pattern: z.string().min(1).optional(),
  valueType: SetValueTypeSchema,
  /** Legal values, for `enum`. */
  values: z.array(z.string().min(1)).optional(),
  /** Legal range, for `number`. */
  range: z.tuple([z.number(), z.number()]).optional(),
  unit: z.string().max(32).optional(),
  appliesTo: SetTargetKindSchema,
  description: z.string().min(1).max(400),
});

/** A registry entry. */
export type SetKeyDecl = z.infer<typeof SetKeyDeclSchema>;

/** The registry itself, so a consumer can validate a hand-edited copy. */
export const SetKeyRegistrySchema = z.array(SetKeyDeclSchema);

const decl = (d: SetKeyDecl): SetKeyDecl => d;

/**
 * Every key the `set` verb accepts.
 *
 * Grouped by namespace: `rules.*` (behaviour switches), `lights.*`, `doors.*`,
 * `pose.*` (articulated state), `signal:*` (traffic control), `env.*`.
 */
export const SET_KEY_REGISTRY: readonly SetKeyDecl[] = [
  decl({
    key: 'rules.collisionAvoidance',
    valueType: 'boolean',
    appliesTo: 'any_actor',
    description:
      'Whether the actor brakes/steers to avoid a collision. Set false to make a challenger commit; without this every generated critical scenario degrades into a safe one.',
  }),
  decl({
    key: 'control:<id>.indication',
    pattern: '^control:[A-Za-z0-9][A-Za-z0-9_.:@/-]{0,127}\\.indication$',
    valueType: 'enum',
    values: ['green', 'yellow', 'red', 'flashing_yellow', 'flashing_red', 'off', 'green_arrow', 'yellow_arrow', 'red_x', 'proceed', 'stop'],
    appliesTo: 'world',
    description: 'Set a portable authored traffic control indication, including work-zone signals, lane-use arrows/X, and human stop/proceed direction.',
  }),
  decl({
    key: 'rules.obeySignals',
    valueType: 'boolean',
    appliesTo: 'any_actor',
    description: 'Whether the actor stops for red lights and stop signs. False models a red-light violator.',
  }),
  decl({
    key: 'rules.obeySpeedLimit',
    valueType: 'boolean',
    appliesTo: 'vehicle',
    description: 'Whether the actor caps its speed at the posted limit.',
  }),
  decl({
    key: 'rules.yieldToVehicles',
    valueType: 'boolean',
    appliesTo: 'any_actor',
    description: 'Whether the actor yields right of way to conflicting vehicles.',
  }),
  decl({
    key: 'rules.yieldToPedestrians',
    valueType: 'boolean',
    appliesTo: 'any_actor',
    description:
      'Whether the actor yields to pedestrians and animals in crossing conflicts. Applies to VRUs too: a non-yielding pedestrian clears this and rules.yieldToVehicles.',
  }),
  decl({
    key: 'rules.aggression',
    valueType: 'number',
    range: [0, 1],
    appliesTo: 'any_actor',
    description: 'Gap acceptance and headway preference, 0 = timid, 1 = aggressive.',
  }),
  decl({
    key: 'rules.laneKeeping',
    valueType: 'boolean',
    appliesTo: 'vehicle',
    description: 'Whether the actor holds its lane centre between explicit lateral interactions.',
  }),
  decl({
    key: 'rules.reactionTimeS',
    valueType: 'number',
    range: [0, 3],
    unit: 's',
    appliesTo: 'any_actor',
    description: 'Perception-reaction delay applied to the actor\'s reactive behaviours.',
  }),

  decl({
    key: 'motion.gear',
    valueType: 'enum',
    values: ['forward', 'reverse'],
    appliesTo: 'vehicle',
    // Kept under the registry's 400-char limit; the rationale for gear-as-discrete-state rather
    // than a negative `speed` target lives in the reverse implementation, not here.
    description:
      'Direction of travel. `reverse` backs the vehicle rear-first along its path, keeping its heading, at a governed low speed. Select it at a standstill; a shift at speed is held until the vehicle stops.',
  }),

  decl({
    key: 'lights.indicator',
    valueType: 'enum',
    values: ['off', 'left', 'right', 'hazard'],
    appliesTo: 'vehicle',
    description: 'Turn-indicator state. An unindicated cut-in is a different scenario from an indicated one.',
  }),
  decl({
    key: 'lights.headlights',
    valueType: 'enum',
    values: ['off', 'drl', 'low', 'high'],
    appliesTo: 'vehicle',
    description: 'Headlight state.',
  }),
  decl({
    key: 'lights.brake',
    valueType: 'boolean',
    appliesTo: 'vehicle',
    description: 'Force brake lights independently of deceleration (phantom-brake bait).',
  }),
  decl({
    key: 'lights.reverse',
    valueType: 'boolean',
    appliesTo: 'vehicle',
    description: 'Reverse lamps.',
  }),
  decl({
    key: 'lights.emergency',
    valueType: 'enum',
    values: ['off', 'flashing', 'flashing_siren'],
    appliesTo: 'vehicle',
    description: 'Emergency-vehicle beacons, optionally with siren.',
  }),
  decl({
    key: 'audio.horn',
    valueType: 'boolean',
    appliesTo: 'vehicle',
    description: 'Vehicle horn state. This is trace-visible and exporters warn when the target profile cannot preserve audio.',
  }),

  decl({
    key: 'doors.left',
    valueType: 'enum',
    values: ['closed', 'opening', 'open', 'closing'],
    appliesTo: 'vehicle',
    description: 'Kerbside/offside left door state — the door-zone scenario for cyclists.',
  }),
  decl({
    key: 'doors.right',
    valueType: 'enum',
    values: ['closed', 'opening', 'open', 'closing'],
    appliesTo: 'vehicle',
    description: 'Right door state.',
  }),
  decl({
    key: 'doors.rear',
    valueType: 'enum',
    values: ['closed', 'opening', 'open', 'closing'],
    appliesTo: 'vehicle',
    description: 'Rear door/tailgate state.',
  }),

  decl({
    key: 'pose.paddle',
    valueType: 'enum',
    values: ['stowed', 'stop', 'slow'],
    appliesTo: 'vru',
    description: 'Flagger paddle state in a work zone.',
  }),
  decl({
    key: 'pose.stopArm',
    valueType: 'enum',
    values: ['retracted', 'extending', 'extended'],
    appliesTo: 'vehicle',
    description: 'School-bus stop arm.',
  }),
  decl({
    key: 'pose.gesture',
    valueType: 'enum',
    values: ['none', 'wave_through', 'halt', 'point', 'phone'],
    appliesTo: 'vru',
    description: 'Upper-body gesture; drives the intent cue a perception stack may or may not read.',
  }),
  decl({
    key: 'pose.headingLookDeg',
    valueType: 'number',
    range: [-180, 180],
    unit: 'deg',
    appliesTo: 'vru',
    description: 'Gaze direction relative to travel direction — the "did they look?" cue.',
  }),

  decl({
    key: 'signal:<handle>.phase',
    pattern: '^signal:[A-Za-z0-9][A-Za-z0-9_.:-]{0,127}\\.phase$',
    valueType: 'enum',
    // `off` is the blackout — a dark head, which reverts the junction to an
    // all-way stop rather than to no control. The arrows are the turn-logic
    // indications; a flashing yellow arrow is what makes a protected left
    // permissive, which is the whole reason the FYA exists.
    values: [
      'green', 'yellow', 'red', 'flashing_yellow', 'flashing_red', 'off',
      'green_arrow', 'yellow_arrow', 'red_x', 'flashing_yellow_arrow', 'flashing_red_arrow',
    ],
    appliesTo: 'world',
    description: 'Force a traffic signal to a phase, including a blackout (`off`). The handle comes from the map catalog.',
  }),
  decl({
    key: 'signal:<handle>.program',
    pattern: '^signal:[A-Za-z0-9][A-Za-z0-9_.:-]{0,127}\\.program$',
    valueType: 'string',
    appliesTo: 'world',
    description: 'Switch a signal to a named program (e.g. night flashing).',
  }),
  decl({
    key: 'signal:feature:<feature>:<approach>.phase',
    pattern: '^signal:feature:[A-Za-z][A-Za-z0-9_-]{0,63}:(subject|ego|opposing|left|right)\\.phase$',
    valueType: 'enum',
    values: [
      'green', 'yellow', 'red', 'flashing_yellow', 'flashing_red', 'off',
      'green_arrow', 'yellow_arrow', 'red_x', 'flashing_yellow_arrow', 'flashing_red_arrow',
    ],
    appliesTo: 'world',
    description: 'Force the signal bound to a portable junction feature and relative approach, including a blackout (`off`).',
  }),
  decl({
    key: 'signal:feature:<feature>:<approach>.program',
    pattern: '^signal:feature:[A-Za-z][A-Za-z0-9_-]{0,63}:(subject|ego|opposing|left|right)\\.program$',
    valueType: 'string',
    appliesTo: 'world',
    description: 'Switch the signal bound to a portable junction feature and relative approach to a named program.',
  }),

  decl({
    key: 'env.weather',
    valueType: 'enum',
    values: [
      'clear',
      'cloudy',
      'overcast',
      'light_rain',
      'heavy_rain',
      'wet_road',
      'fog_light',
      'fog_dense',
      'snow',
      'sleet',
    ],
    appliesTo: 'world',
    description: 'Change weather mid-clip (a squall arriving), overriding `environment.weather`.',
  }),
  decl({
    key: 'env.frictionScale',
    valueType: 'number',
    range: [0.1, 1.2],
    appliesTo: 'world',
    description: 'Tyre-road friction multiplier; drives the friction-patch scenarios.',
  }),
  decl({
    key: 'env.fogDensity',
    valueType: 'number',
    range: [0, 1],
    appliesTo: 'world',
    description: 'Fog density, 0 = clear.',
  }),
  decl({
    key: 'env.rainIntensity',
    valueType: 'number',
    range: [0, 1],
    appliesTo: 'world',
    description: 'Rain intensity, 0 = dry.',
  }),
];

const LITERAL_KEYS = new Map(
  SET_KEY_REGISTRY.filter((d) => !d.pattern).map((d) => [d.key, d] as const),
);
const PATTERN_KEYS = SET_KEY_REGISTRY.filter((d) => d.pattern).map(
  (d) => [new RegExp(d.pattern as string), d] as const,
);

/** Look up a key in the registry, honouring wildcard patterns. */
export function lookupSetKey(key: string): SetKeyDecl | undefined {
  const literal = LITERAL_KEYS.get(key);
  if (literal) return literal;
  for (const [pattern, d] of PATTERN_KEYS) {
    if (pattern.test(key)) return d;
  }
  return undefined;
}

/** Namespace of a key (`rules`, `lights`, `doors`, `pose`, `signal`, `env`). */
export function setKeyNamespace(key: string): string {
  if (key.startsWith('signal:')) return 'signal';
  if (key.startsWith('control:')) return 'control';
  const dot = key.indexOf('.');
  return dot < 0 ? key : key.slice(0, dot);
}

/** Every literal key, sorted — for error messages and editor autocomplete. */
export function knownSetKeys(): string[] {
  return SET_KEY_REGISTRY.map((d) => d.key).sort();
}

/** Cheap edit-distance suggestion for a mistyped key. */
export function suggestSetKey(key: string): string | undefined {
  const candidates = knownSetKeys();
  let best: { key: string; score: number } | undefined;
  for (const candidate of candidates) {
    const score = levenshtein(key.toLowerCase(), candidate.toLowerCase());
    if (!best || score < best.score) best = { key: candidate, score };
  }
  return best && best.score <= Math.max(3, Math.floor(key.length / 3)) ? best.key : undefined;
}

function levenshtein(a: string, b: string): number {
  const prev = new Array<number>(b.length + 1);
  const cur = new Array<number>(b.length + 1);
  for (let j = 0; j <= b.length; j++) prev[j] = j;
  for (let i = 1; i <= a.length; i++) {
    cur[0] = i;
    for (let j = 1; j <= b.length; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      cur[j] = Math.min(
        (cur[j - 1] as number) + 1,
        (prev[j] as number) + 1,
        (prev[j - 1] as number) + cost,
      );
    }
    for (let j = 0; j <= b.length; j++) prev[j] = cur[j] as number;
  }
  return prev[b.length] as number;
}

/** The value shapes `set` accepts. */
export const SetValueSchema = z.union([z.boolean(), z.number().finite(), z.string().max(200)]);
/** A `set` value. */
export type SetValue = z.infer<typeof SetValueSchema>;

/** Why a `set` value was rejected. `ok` carries no reason. */
export type SetValueCheck =
  | { ok: true }
  | { ok: false; code: 'unknown_set_key' | 'set_value_type' | 'set_value_range'; message: string };

/** Validate a `key`/`value` pair against the registry. */
export function checkSetValue(key: string, value: SetValue): SetValueCheck {
  const declared = lookupSetKey(key);
  if (!declared) {
    const suggestion = suggestSetKey(key);
    return {
      ok: false,
      code: 'unknown_set_key',
      message: `unknown set key "${key}"${suggestion ? `; did you mean "${suggestion}"?` : ''}`,
    };
  }
  switch (declared.valueType) {
    case 'boolean':
      return typeof value === 'boolean'
        ? { ok: true }
        : { ok: false, code: 'set_value_type', message: `"${key}" takes a boolean` };
    case 'number': {
      if (typeof value !== 'number') {
        return { ok: false, code: 'set_value_type', message: `"${key}" takes a number` };
      }
      if (declared.range && (value < declared.range[0] || value > declared.range[1])) {
        return {
          ok: false,
          code: 'set_value_range',
          message: `"${key}" must be within [${declared.range[0]}, ${declared.range[1]}], got ${value}`,
        };
      }
      return { ok: true };
    }
    case 'enum': {
      if (typeof value !== 'string' || !declared.values?.includes(value)) {
        return {
          ok: false,
          code: 'set_value_type',
          message: `"${key}" takes one of: ${declared.values?.join(', ')}`,
        };
      }
      return { ok: true };
    }
    case 'string':
      return typeof value === 'string'
        ? { ok: true }
        : { ok: false, code: 'set_value_type', message: `"${key}" takes a string` };
  }
}

File: native/crates/simforge-authoring/src/validate/set_keys.rs
//! The typed key registry behind the `set` verb.
//!
//! `set` is the discrete-state escape valve (indicators, doors, a flagger's
//! paddle, a signal phase, behaviour switches). Left open it becomes a
//! stringly-typed channel no engine can implement completely, so the legal
//! keys are data, each with a value type and the actor kinds it applies to,
//! and anything else is refused with the near-miss suggestion. Wildcard keys
//! (`signal:<handle>.phase`) match by pattern, since the handles come from the
//! map. Kept entry for entry with the scenario schema's registry.

/// Value domain of a settable key.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ValueType {
    Boolean,
    Number,
    Enum,
    String,
}

/// Who a key may be set on.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AppliesTo {
    Vehicle,
    Vru,
    AnyActor,
    World,
}

impl AppliesTo {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Vehicle => "vehicle",
            Self::Vru => "vru",
            Self::AnyActor => "any_actor",
            Self::World => "world",
        }
    }
}

/// How a wildcard key is recognised (the registry's regex patterns, spelled
/// out: a prefix, a body drawn from a character class, and a suffix).
#[derive(Debug, Clone, Copy)]
enum Pattern {
    /// `^control:[A-Za-z0-9][A-Za-z0-9_.:@/-]{0,127}\.indication$`
    Control,
    /// `^signal:[A-Za-z0-9][A-Za-z0-9_.:-]{0,127}\.<suffix>$`
    SignalHandle(&'static str),
    /// `^signal:feature:[A-Za-z][A-Za-z0-9_-]{0,63}:(subject|ego|opposing|left|right)\.<suffix>$`
    SignalFeature(&'static str),
}

impl Pattern {
    fn test(self, key: &str) -> bool {
        match self {
            Self::Control => key
                .strip_prefix("control:")
                .and_then(|rest| rest.strip_suffix(".indication"))
                .is_some_and(|body| {
                    handle_body(body, |b| {
                        b.is_ascii_alphanumeric() || b"_.:@/-".contains(&b)
                    })
                }),
            Self::SignalHandle(suffix) => key
                .strip_prefix("signal:")
                .and_then(|rest| rest.strip_suffix(suffix))
                .and_then(|rest| rest.strip_suffix('.'))
                .is_some_and(|body| {
                    handle_body(body, |b| b.is_ascii_alphanumeric() || b"_.:-".contains(&b))
                }),
            Self::SignalFeature(suffix) => key
                .strip_prefix("signal:feature:")
                .and_then(|rest| rest.strip_suffix(suffix))
                .and_then(|rest| rest.strip_suffix('.'))
                .and_then(|rest| rest.rsplit_once(':'))
                .is_some_and(|(feature, approach)| {
                    matches!(approach, "subject" | "ego" | "opposing" | "left" | "right")
                        && feature.len() <= 64
                        && feature
                            .bytes()
                            .next()
                            .is_some_and(|b| b.is_ascii_alphabetic())
                        && feature
                            .bytes()
                            .all(|b| b.is_ascii_alphanumeric() || b == b'_' || b == b'-')
                }),
        }
    }
}

/// `[A-Za-z0-9]<rest class>{0,127}`.
fn handle_body(body: &str, rest: impl Fn(u8) -> bool) -> bool {
    let bytes = body.as_bytes();
    !bytes.is_empty()
        && bytes.len() <= 128
        && bytes[0].is_ascii_alphanumeric()
        && bytes[1..].iter().all(|&b| rest(b))
}

/// One registry entry.
#[derive(Debug, Clone, Copy)]
pub struct SetKeyDecl {
    pub key: &'static str,
    pattern: Option<Pattern>,
    pub value_type: ValueType,
    pub values: &'static [&'static str],
    pub range: Option<(f64, f64)>,
    pub applies_to: AppliesTo,
}

const fn lit(key: &'static str, value_type: ValueType, applies_to: AppliesTo) -> SetKeyDecl {
    SetKeyDecl {
        key,
        pattern: None,
        value_type,
        values: &[],
        range: None,
        applies_to,
    }
}

const fn enm(
    key: &'static str,
    values: &'static [&'static str],
    applies_to: AppliesTo,
) -> SetKeyDecl {
    SetKeyDecl {
        key,
        pattern: None,
        value_type: ValueType::Enum,
        values,
        range: None,
        applies_to,
    }
}

const fn num(key: &'static str, lo: f64, hi: f64, applies_to: AppliesTo) -> SetKeyDecl {
    SetKeyDecl {
        key,
        pattern: None,
        value_type: ValueType::Number,
        values: &[],
        range: Some((lo, hi)),
        applies_to,
    }
}

const fn pat(
    key: &'static str,
    pattern: Pattern,
    value_type: ValueType,
    values: &'static [&'static str],
) -> SetKeyDecl {
    SetKeyDecl {
        key,
        pattern: Some(pattern),
        value_type,
        values,
        range: None,
        applies_to: AppliesTo::World,
    }
}

const SIGNAL_PHASES: &[&str] = &[
    "green",
    "yellow",
    "red",
    "flashing_yellow",
    "flashing_red",
    "off",
    "green_arrow",
    "yellow_arrow",
    "red_x",
    "flashing_yellow_arrow",
    "flashing_red_arrow",
];

const DOOR_STATES: &[&str] = &["closed", "opening", "open", "closing"];

use AppliesTo::{AnyActor, Vehicle, Vru, World};
use ValueType::{Boolean, String as Str};

/// Every key the `set` verb accepts, in registry order.
pub const SET_KEY_REGISTRY: &[SetKeyDecl] = &[
    lit("rules.collisionAvoidance", Boolean, AnyActor),
    pat(
        "control:<id>.indication",
        Pattern::Control,
        ValueType::Enum,
        &[
            "green",
            "yellow",
            "red",
            "flashing_yellow",
            "flashing_red",
            "off",
            "green_arrow",
            "yellow_arrow",
            "red_x",
            "proceed",
            "stop",
        ],
    ),
    lit("rules.obeySignals", Boolean, AnyActor),
    lit("rules.obeySpeedLimit", Boolean, Vehicle),
    lit("rules.yieldToVehicles", Boolean, AnyActor),
    lit("rules.yieldToPedestrians", Boolean, AnyActor),
    num("rules.aggression", 0.0, 1.0, AnyActor),
    lit("rules.laneKeeping", Boolean, Vehicle),
    num("rules.reactionTimeS", 0.0, 3.0, AnyActor),
    enm("motion.gear", &["forward", "reverse"], Vehicle),
    enm(
        "lights.indicator",
        &["off", "left", "right", "hazard"],
        Vehicle,
    ),
    enm("lights.headlights", &["off", "drl", "low", "high"], Vehicle),
    lit("lights.brake", Boolean, Vehicle),
    lit("lights.reverse", Boolean, Vehicle),
    enm(
        "lights.emergency",
        &["off", "flashing", "flashing_siren"],
        Vehicle,
    ),
    lit("audio.horn", Boolean, Vehicle),
    enm("doors.left", DOOR_STATES, Vehicle),
    enm("doors.right", DOOR_STATES, Vehicle),
    enm("doors.rear", DOOR_STATES, Vehicle),
    enm("pose.paddle", &["stowed", "stop", "slow"], Vru),
    enm(
        "pose.stopArm",
        &["retracted", "extending", "extended"],
        Vehicle,
    ),
    enm(
        "pose.gesture",
        &["none", "wave_through", "halt", "point", "phone"],
        Vru,
    ),
    num("pose.headingLookDeg", -180.0, 180.0, Vru),
    pat(
        "signal:<handle>.phase",
        Pattern::SignalHandle("phase"),
        ValueType::Enum,
        SIGNAL_PHASES,
    ),
    pat(
        "signal:<handle>.program",
        Pattern::SignalHandle("program"),
        Str,
        &[],
    ),
    pat(
        "signal:feature:<feature>:<approach>.phase",
        Pattern::SignalFeature("phase"),
        ValueType::Enum,
        SIGNAL_PHASES,
    ),
    pat(
        "signal:feature:<feature>:<approach>.program",
        Pattern::SignalFeature("program"),
        Str,
        &[],
    ),
    enm(
        "env.weather",
        &[
            "clear",
            "cloudy",
            "overcast",
            "light_rain",
            "heavy_rain",
            "wet_road",
            "fog_light",
            "fog_dense",
            "snow",
            "sleet",
        ],
        World,
    ),
    num("env.frictionScale", 0.1, 1.2, World),
    num("env.fogDensity", 0.0, 1.0, World),
    num("env.rainIntensity", 0.0, 1.0, World),
];

/// Look a key up, literal keys first, then the wildcard patterns in order.
pub fn lookup_set_key(key: &str) -> Option<&'static SetKeyDecl> {
    SET_KEY_REGISTRY
        .iter()
        .find(|d| d.pattern.is_none() && d.key == key)
        .or_else(|| {
            SET_KEY_REGISTRY
                .iter()
                .find(|d| d.pattern.is_some_and(|p| p.test(key)))
        })
}

/// Namespace of a key (`rules`, `lights`, `doors`, `pose`, `signal`, `control`, `env`).
pub fn set_key_namespace(key: &str) -> &str {
    if key.starts_with("signal:") {
        return "signal";
    }
    if key.starts_with("control:") {
        return "control";
    }
    key.split_once('.').map_or(key, |(ns, _)| ns)
}

/// Every key, sorted (JavaScript default sort).
fn known_set_keys() -> Vec<&'static str> {
    let mut keys: Vec<&'static str> = SET_KEY_REGISTRY.iter().map(|d| d.key).collect();
    keys.sort_by(|a, b| simforge_core::hash::cmp_utf16(a, b));
    keys
}

/// Cheap edit-distance suggestion for a mistyped key.
fn suggest_set_key(key: &str) -> Option<&'static str> {
    let lowered: Vec<u16> = key.to_lowercase().encode_utf16().collect();
    let mut best: Option<(&'static str, usize)> = None;
    for candidate in known_set_keys() {
        let c: Vec<u16> = candidate.to_lowercase().encode_utf16().collect();
        let score = levenshtein(&lowered, &c);
        if best.is_none_or(|(_, s)| score < s) {
            best = Some((candidate, score));
        }
    }
    let limit = 3usize.max(key.encode_utf16().count() / 3);
    best.filter(|(_, score)| *score <= limit).map(|(k, _)| k)
}

fn levenshtein(a: &[u16], b: &[u16]) -> usize {
    let mut prev: Vec<usize> = (0..=b.len()).collect();
    let mut cur = vec![0usize; b.len() + 1];
    for i in 1..=a.len() {
        cur[0] = i;
        for j in 1..=b.len() {
            let cost = usize::from(a[i - 1] != b[j - 1]);
            cur[j] = (cur[j - 1] + 1).min(prev[j] + 1).min(prev[j - 1] + cost);
        }
        prev.clone_from(&cur);
    }
    prev[b.len()]
}

/// Why a `set` value was refused: `(code, message)`.
pub fn check_set_value(
    key: &str,
    value: &simforge_compiler::template::SetValue,
) -> Result<(), (&'static str, String)> {
    use crate::jsfmt::js;
    use simforge_compiler::template::SetValue;
    let Some(declared) = lookup_set_key(key) else {
        let suggestion = suggest_set_key(key)
            .map(|s| format!("; did you mean \"{s}\"?"))
            .unwrap_or_default();
        return Err((
            "unknown_set_key",
            format!("unknown set key \"{key}\"{suggestion}"),
        ));
    };
    match declared.value_type {
        ValueType::Boolean => match value {
            SetValue::Bool(_) => Ok(()),
            _ => Err(("set_value_type", format!("\"{key}\" takes a boolean"))),
        },
        ValueType::Number => {
            let SetValue::Number(n) = value else {
                return Err(("set_value_type", format!("\"{key}\" takes a number")));
            };
            if let Some((lo, hi)) = declared.range {
                if *n < lo || *n > hi {
                    return Err((
                        "set_value_range",
                        format!(
                            "\"{key}\" must be within [{}, {}], got {}",
                            js(lo),
                            js(hi),
                            js(*n)
                        ),
                    ));
                }
            }
            Ok(())
        }
        ValueType::Enum => match value {
            SetValue::Text(s) if declared.values.contains(&s.as_str()) => Ok(()),
            _ => Err((
                "set_value_type",
                format!("\"{key}\" takes one of: {}", declared.values.join(", ")),
            )),
        },
        ValueType::String => match value {
            SetValue::Text(_) => Ok(()),
            _ => Err(("set_value_type", format!("\"{key}\" takes a string"))),
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn wildcards_match_like_the_registry_patterns() {
        assert_eq!(
            lookup_set_key("signal:J12.3.phase").unwrap().key,
            "signal:<handle>.phase"
        );
        assert_eq!(
            lookup_set_key("signal:J12.program").unwrap().key,
            "signal:<handle>.program"
        );
        assert_eq!(
            lookup_set_key("control:work-zone@a/b.indication")
                .unwrap()
                .key,
            "control:<id>.indication"
        );
        assert!(lookup_set_key("control:.indication").is_none());
        assert!(lookup_set_key("signal:.phase").is_none());
        assert_eq!(
            lookup_set_key("rules.aggression").unwrap().key,
            "rules.aggression"
        );
        assert!(lookup_set_key("rules.agression").is_none());
    }

    #[test]
    fn suggestions() {
        assert_eq!(suggest_set_key("rules.agression"), Some("rules.aggression"));
        assert_eq!(
            suggest_set_key("zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz"),
            None
        );
        assert_eq!(set_key_namespace("lights.brake"), "lights");
        assert_eq!(set_key_namespace("signal:x.phase"), "signal");
        assert_eq!(set_key_namespace("nodot"), "nodot");
    }
}
